### PR TITLE
Add access token requesting [#168661200]

### PR DESCRIPTION
**Pull request by vitaliypodlypnyi**
1. Added functionality to request access tokens (using auth code and refresh tokens)
2. Added JS to handle iframe notifications and send JWTs to iframe
3. Updated 401 exception handling

### CODE REVIEW

File: CredentialPostTest/CredentialPostTest/Data/Models/ZwapstoreRequest.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+namespace CredentialPostTest.Data.Models
+{
+    public class ZwapstoreRequest
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/Data/Models/ZwapstoreRequest.cs:3]: Rename to `ClientRequest`?

File: CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs [hunk 2]
```diff
@@ -193,50 +235,55 @@ private async Task<string> GetOneTimeCodeAsync()
...
-                {
-                    Content = new StringContent(JsonConvert.SerializeObject(requestModel), Encoding.UTF8, "application/json")
-                })
-                {
-                    var response = await GetResponseMessageAsync(request, restClient);
-
-                    responseContent = await response.Content.ReadAsStringAsync();
-                }
+                var content = new StringContent(JsonConvert.SerializeObject(requestModel), Encoding.UTF8, "application/json");
+                var response = await GetResponseMessageAsync(HttpMethod.Post, $"{_zgApiUrl}api/zwapstore/{endpoint}", content, restClient, withAuthorization);
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs:284]: Why are we using an internal endpoint here? Plz fix so it uses public API.

File: CredentialPostTest/CredentialPostTest/wwwroot/js/site.js [hunk 3]
```diff
@@ -35,9 +58,25 @@ function processIntegrationCreatedResult(systems) {
...
     if (message) {
         message.removeAttribute("hidden");
         message.innerHTML = 'Integration between <b>' + sourceSystem + '</b> and <b>' + targetSystem + '</b> was created successfully!';
     }
+}
+
+function processClientAuthenticatedMessage(authMessage){
+    let code = authMessage.authorizationCode;
+    callApi('GET', '/Index?handler=AccessToken&authCode=' + code, function(data){
+        let accToken = data.accessToken;
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/wwwroot/js/site.js:71]: Access token in plain text?!

File: CredentialPostTest/CredentialPostTest/wwwroot/js/site.js [hunk 4]
```diff
@@ -35,9 +58,25 @@ function processIntegrationCreatedResult(systems) {
...
+        let otc = data.otc;
+        sendToIframe('accessToken', accToken);
+    })
+}
+
+function processAccessTokenMissingMessage(){
+    callApi('GET', '/Index?handler=RefreshAccessToken', function(data){
+        let accToken = data.accessToken;
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/wwwroot/js/site.js:79]: Access token in plain text?!

**Review by dennisahlin — CHANGES_REQUESTED** [review]
See comments

File: CredentialPostTest/CredentialPostTest/Data/Models/ZwapstoreRequest.cs (hunk 1 above)
Comment (KaterynaZorina) [reply CredentialPostTest/CredentialPostTest/Data/Models/ZwapstoreRequest.cs:3]: Maybe `public abstract class ZwapstoreRequestBase`? As I understand, this class was added in order to be used as a base class, right?

File: CredentialPostTest/CredentialPostTest/wwwroot/js/site.js (hunk 3 above)
Comment (KaterynaZorina) [reply CredentialPostTest/CredentialPostTest/wwwroot/js/site.js:71]: We can return encrypted access token to client from our API.

**Review by KaterynaZorina — CHANGES_REQUESTED** [review]

File: CredentialPostTest/CredentialPostTest/wwwroot/js/site.js (hunk 3 above)
Comment (dennisahlin) [reply CredentialPostTest/CredentialPostTest/wwwroot/js/site.js:71]: Or we can pass the authorizationCode to window so exchange can be done serverside?

File: CredentialPostTest/CredentialPostTest/Data/Models/OauthRequest.cs [hunk 5]
```diff
@@ -1,11 +1,17 @@
 namespace CredentialPostTest.Data.Models
 {
-    internal class RefreshTokenRequest
+    internal class OauthRequest
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/Data/Models/OauthRequest.cs:3]: `Oauth` => `Oauth2`

File: CredentialPostTest/CredentialPostTest/Data/Models/OauthRequest.cs [hunk 6]
```diff
@@ -1,11 +1,17 @@
...
     {
         public string ClientId { get; set; }
 
         public string ClientSecret { get; set; }
-
+
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/Data/Models/OauthRequest.cs:8]: Extra whitespace?

File: CredentialPostTest/CredentialPostTest/Data/Models/OauthResponse.cs [hunk 7]
```diff
@@ -1,6 +1,11 @@
 namespace CredentialPostTest.Data.Models
 {
-    internal class RefreshTokenResponse
+    internal class OauthResponse
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/Data/Models/OauthResponse.cs:3]: `Oauth` => `Oauth2`

File: CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs [hunk 8]
```diff
@@ -81,12 +84,62 @@ public async Task OnGet()
...
+        public async Task<IActionResult> OnGetAccessToken(string authCode)
+        {
+            var request = new OauthRequest
+            {
+                ClientId = _clientId,
+                ClientSecret = _clientSecret,
+                Code = authCode,
+                GrantType = "authorization_code"
+            };
+            var res = await Post<OauthRequest, OauthResponse>(request, "oauth", "token", false);
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs:99]: `oauth` => `oauth2`

File: CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs [hunk 9]
```diff
@@ -81,12 +84,62 @@ public async Task OnGet()
...
+            var user = await _userManager.GetUserAsync(User);
+            var request = new OauthRequest
+            {
+                ClientId = _clientId,
+                ClientSecret = _clientSecret,
+                RefreshToken = user.RefreshToken,
+                GrantType = "refresh_token"
+            };
+            
+            var res = await Post<OauthRequest, OauthResponse>(request, "oauth", "token", false);
```
Comment (dennisahlin) [inline CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs:130]: `oauth` => `oauth2`

Comment (dennisahlin) [reply CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs:130]: Use full words please, `res` => `result`

File: CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs (hunk 8 above)
Comment (dennisahlin) [reply CredentialPostTest/CredentialPostTest/Pages/Index.cshtml.cs:99]: `res` => `result`

**Review by dennisahlin — CHANGES_REQUESTED** [review]

**Review by KaterynaZorina — APPROVED** [review]

**Comment by dennisahlin** [comment]
Now, as a last check. Put yourself into a devs position in some other company and consider that you're tasked to implement Zwapstore into your service and you're handed this example. Would you understand how you should apply it from this example? Is anything missing? Are functions properly and informatively named, is there comments to explain why things are done in proper places etc. etc.? This example project should be helpful and clear for any external part that wants to implement this.

**Comment by vitaliypodlypnyi** [comment]
got it, adding comments

**Review by dennisahlin — APPROVED** [review]

**Comment by dennisahlin** [comment]
Hm, maybe not merge it yet, since we haven't released the API?

**Comment by vitaliypodlypnyi** [comment]
already merged, but we can revert all related commits

**Comment by dennisahlin** [comment]
Meh, no-one is looking at it atm anyways.

**Pull request merged**